`StreamingOutput` blocking API does not call `reset()` on error - clients cannot detect incomplete chunked responses

### Describe the bug

Reviewing and doing a small regression test development coverage for issue #50336 , I've noticed that the fix in [PR #50361 ](https://github.com/quarkusio/quarkus/pull/50361) only addressed reactive streaming (`Multi`/`Publisher`) but did not cover the blocking `StreamingOutput` API.


### Expected behavior

Similar to the fix (PR #50361 )for reactive streams (`Multi`), when an exception occurs during `StreamingOutput` streaming:
- The server should call `HttpServerResponse.reset()` 
- The final zero-length chunk should NOT be sent
- HTTP clients should be able to detect the connection was abruptly closed

### Actual behavior

When using `StreamingOutput` with `@Blocking`, if an exception occurs during streaming:
- The chunked response is closed normally
- HTTP clients cannot detect the incomplete transfer

This is the error trace from my test:

```
2025-10-28 08:20:42,579 ERROR [io.ver.cor.htt.imp.HttpClientRequestImpl] (vert.x-eventloop-thread-1) Connection was closed
08:20:43,588 INFO  [app] 08:20:42,485 HTTP Request to /streaming-output-error?fail=true failed, error id: 253e42d7-48b1-40ab-8b7b-04606bc70b9a-1: java.lang.RuntimeException: java.io.IOException: dummy failure
08:20:43,589 INFO  [app]        at io.quarkus.ts.vertx.StreamingOutputErrorResource.lambda$stream$0(StreamingOutputErrorResource.java:41)
08:20:43,589 INFO  [app]        at org.jboss.resteasy.reactive.server.providers.serialisers.StreamingOutputMessageBodyWriter.writeResponse(StreamingOutputMessageBodyWriter.java:49)
08:20:43,590 INFO  [app]        at org.jboss.resteasy.reactive.server.providers.serialisers.StreamingOutputMessageBodyWriter.writeResponse(StreamingOutputMessageBodyWriter.java:17)
08:20:43,590 INFO  [app]        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:218)
08:20:43,591 INFO  [app]        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:186)
08:20:43,591 INFO  [app]        at org.jboss.resteasy.reactive.server.core.serialization.FixedEntityWriterArray.write(FixedEntityWriterArray.java:31)
08:20:43,592 INFO  [app]        at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:34)
08:20:43,593 INFO  [app]        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:189)
08:20:43,593 INFO  [app]        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
08:20:43,593 INFO  [app]        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:645)
08:20:43,594 INFO  [app]        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
08:20:43,594 INFO  [app]        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
08:20:43,595 INFO  [app]        at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
08:20:43,595 INFO  [app]        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
08:20:43,596 INFO  [app]        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
08:20:43,596 INFO  [app]        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
08:20:43,597 INFO  [app]        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
08:20:43,597 INFO  [app]        at java.base/java.lang.Thread.run(Thread.java:1583)
08:20:43,598 INFO  [app] Caused by: java.io.IOException: dummy failure
08:20:43,598 INFO  [app]        at io.quarkus.ts.vertx.StreamingOutputErrorResource.lambda$stream$0(StreamingOutputErrorResource.java:35)
08:20:43,599 INFO  [app]        ... 17 more
08:20:43,599 INFO  [app] 08:20:42,543 Exception in SSE server handling, impossible to send it to client: io.vertx.core.impl.NoStackTraceException: dummy failure
08:20:44,700 INFO  [app] Service stopped (Quarkus JVM mode)
[ERROR] Tests run: 4, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.077 s <<< FAILURE! -- in io.quarkus.ts.vertx.StreamingErrorIT
[ERROR] io.quarkus.ts.vertx.StreamingErrorIT.testStreamingOutputFailureMidStream -- Time elapsed: 0.022 s <<< FAILURE!
org.opentest4j.AssertionFailedError: Client should have failed as the server reset the connection (StreamingOutput) ==> Expected java.util.concurrent.ExecutionException to be thrown, but nothing was thrown.
        at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:152)
        at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:73)
        at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:39)
        at org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3153)
        at io.quarkus.ts.vertx.StreamingErrorIT.lambda$testStreamingOutputFailureMidStream$9(StreamingErrorIT.java:114)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   StreamingErrorIT.lambda$testStreamingOutputFailureMidStream$9:114 Client should have failed as the server reset the connection (StreamingOutput) ==> Expected java.util.concurrent.ExecutionException to be thrown, but nothing was thrown.
[INFO] 
```



### How to Reproduce?

Go to my branch https://github.com/jcarranzan/quarkus-test-suite/blob/test-development/chunked-response-issue-50336/http/vertx/src/test/java/io/quarkus/ts/vertx/StreamingErrorIT.java  here and execute the test `testStreamingOutputFailureMidStream`

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/providers/serialisers/StreamingOutputMessageBodyWriter.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/providers/serialisers/StreamingOutputMessageBodyWriter.java
@@ -10,11 +10,13 @@ import jakarta.ws.rs.core.MediaType;
 import jakarta.ws.rs.core.MultivaluedMap;
 import jakarta.ws.rs.core.StreamingOutput;
 
+import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
 import org.jboss.resteasy.reactive.server.spi.ResteasyReactiveResourceInfo;
 import org.jboss.resteasy.reactive.server.spi.ServerMessageBodyWriter;
 import org.jboss.resteasy.reactive.server.spi.ServerRequestContext;
 
 public class StreamingOutputMessageBodyWriter implements ServerMessageBodyWriter<StreamingOutput> {
+
     @Override
     public boolean isWriteable(Class<?> type, Type genericType, Annotation[] annotations, MediaType mediaType) {
         return doIsWriteable(type);
@@ -45,7 +47,25 @@ public class StreamingOutputMessageBodyWriter implements ServerMessageBodyWriter
 
     @Override
     public void writeResponse(StreamingOutput o, Type genericType, ServerRequestContext context)
-            throws WebApplicationException, IOException {
-        o.write(context.getOrCreateOutputStream());
+            throws WebApplicationException {
+        ResteasyReactiveRequestContext rrContext = (ResteasyReactiveRequestContext) context;
+        try {
+            o.write(context.getOrCreateOutputStream());
+        } catch (Throwable t) {
+            if (context.serverResponse().headWritten()) {
+                context.serverResponse().reset();
+                rrContext.resume(t);
+            } else {
+                if (t instanceof WebApplicationException) {
+                    throw (WebApplicationException) t;
+                } else if (t instanceof IOException) {
+                    throw new WebApplicationException(t);
+                } else if (t instanceof RuntimeException) {
+                    throw new WebApplicationException(t);
+                } else {
+                    throw new WebApplicationException(t);
+                }
+            }
+        }
     }
 }

--- a/integration-tests/elytron-resteasy-reactive/src/main/java/io/quarkus/it/resteasy/reactive/elytron/StreamingOutputResource.java
+++ b/integration-tests/elytron-resteasy-reactive/src/main/java/io/quarkus/it/resteasy/reactive/elytron/StreamingOutputResource.java
@@ -1,0 +1,47 @@
+package io.quarkus.it.resteasy.reactive.elytron;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+
+import jakarta.ws.rs.DefaultValue;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.StreamingOutput;
+
+import io.smallrye.common.annotation.Blocking;
+
+@Path("/streaming-output-error")
+public class StreamingOutputResource {
+    private static final int ITEMS_PER_EMIT = 100;
+
+    private static final byte[] CHUNK = "This is one chunk of data.\n".getBytes(StandardCharsets.UTF_8);
+
+    @GET
+    @Path("/output")
+    @Produces(MediaType.TEXT_PLAIN)
+    @Blocking
+    public StreamingOutput streamOutput(@QueryParam("fail") @DefaultValue("false") boolean fail) {
+        return outputStream -> {
+            try {
+                writeData(outputStream);
+                if (fail) {
+                    throw new IOException("dummy failure");
+                }
+                writeData(outputStream);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        };
+    }
+
+    private void writeData(OutputStream out) throws IOException {
+        for (int i = 0; i < ITEMS_PER_EMIT; i++) {
+            out.write(CHUNK);
+            out.flush();
+        }
+    }
+}

--- a/integration-tests/elytron-resteasy-reactive/src/test/java/io/quarkus/it/resteasy/reactive/elytron/StreamingOutputErrorHandlingTest.java
+++ b/integration-tests/elytron-resteasy-reactive/src/test/java/io/quarkus/it/resteasy/reactive/elytron/StreamingOutputErrorHandlingTest.java
@@ -1,0 +1,115 @@
+package io.quarkus.it.resteasy.reactive.elytron;
+
+import static io.restassured.RestAssured.port;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpClient;
+import io.vertx.core.http.HttpClosedException;
+import io.vertx.core.http.HttpMethod;
+
+@QuarkusTest
+public class StreamingOutputErrorHandlingTest {
+    private static final Duration TIMEOUT = Duration.ofSeconds(10);
+    private static final int ITEMS_PER_BATCH = 100;
+    private static final int BYTES_PER_CHUNK = "This is one chunk of data.\n"
+            .getBytes(StandardCharsets.UTF_8).length;
+    private static final long EXPECTED_BYTES_FIRST_BATCH = (long) ITEMS_PER_BATCH * BYTES_PER_CHUNK;
+    private static final long EXPECTED_BYTES_COMPLETE = (long) ITEMS_PER_BATCH * 2 * BYTES_PER_CHUNK;
+
+    private Vertx vertx;
+    private HttpClient client;
+
+    @BeforeEach
+    public void setup() {
+        vertx = Vertx.vertx();
+        client = vertx.createHttpClient();
+    }
+
+    @AfterEach
+    public void cleanup() throws Exception {
+        if (client != null) {
+            client.close().toCompletionStage().toCompletableFuture().get(5, TimeUnit.SECONDS);
+        }
+        if (vertx != null) {
+            vertx.close().toCompletionStage().toCompletableFuture().get(5, TimeUnit.SECONDS);
+        }
+    }
+
+    @Test
+    public void testStreamingOutputFailureMidStream() {
+        AtomicLong byteCount = new AtomicLong();
+        CompletableFuture<Void> latch = new CompletableFuture<>();
+
+        sendRequest("/streaming-output-error/output?fail=true", latch,
+                b -> byteCount.addAndGet(b.length()));
+
+        Assertions.assertTimeoutPreemptively(TIMEOUT, () -> {
+            ExecutionException ex = Assertions.assertThrows(ExecutionException.class,
+                    latch::get,
+                    "Client should have detected that the server reset the connection");
+
+            Assertions.assertInstanceOf(HttpClosedException.class, ex.getCause(),
+                    "Expected HttpClosedException when connection is reset mid-stream");
+        });
+
+        Assertions.assertEquals(EXPECTED_BYTES_FIRST_BATCH, byteCount.get(),
+                "Should have received only the first batch of data before failure");
+    }
+
+    @Test
+    public void testStreamingOutputSuccess() {
+        AtomicLong byteCount = new AtomicLong();
+        CompletableFuture<Void> latch = new CompletableFuture<>();
+
+        sendRequest("/streaming-output-error/output?fail=false", latch,
+                b -> byteCount.addAndGet(b.length()));
+
+        Assertions.assertTimeoutPreemptively(TIMEOUT,
+                () -> latch.get(),
+                "StreamingOutput should complete successfully without errors");
+
+        Assertions.assertEquals(EXPECTED_BYTES_COMPLETE, byteCount.get(),
+                "Should have received all data when no errors occur");
+    }
+
+    private void sendRequest(String uri, CompletableFuture<Void> latch, Consumer<Buffer> bodyConsumer) {
+        Handler<Throwable> failureHandler = latch::completeExceptionally;
+
+        client.request(HttpMethod.GET, port, "localhost", uri)
+                .onFailure(failureHandler)
+                .onSuccess(request -> {
+                    request.end();
+                    request.connect()
+                            .onFailure(failureHandler)
+                            .onSuccess(response -> {
+                                response.request().connection().closeHandler(v -> {
+                                    failureHandler.handle(new HttpClosedException("Connection was closed"));
+                                });
+
+                                response.handler(buffer -> {
+                                    if (buffer.length() > 0) {
+                                        bodyConsumer.accept(buffer);
+                                    }
+                                });
+                                response.exceptionHandler(failureHandler);
+                                response.endHandler(v -> latch.complete(null));
+                            });
+                });
+    }
+}

